qiita_db/software.py

```diff
@@ -0,0 +1,522 @@
+            the software
+        """
+        with TRN:
+            sql = """INSERT INTO qiita.software (name, version, description)
+                        VALUES (%s, %s, %s)
+                        RETURNING software_id"""
+            sql_params = [name, version, description]
+            TRN.add(sql, sql_params)
+            s_id = TRN.execute_fetchlast()
+
+            instance = cls(s_id)
+
+            if publications:
+                instance.add_publications(publications)
+
+        return instance
+
+    @property
+    def name(self):
+        """The name of the software
+
+        Returns
+        -------
+        str
+            The name of the software
+        """
+        with TRN:
+            sql = "SELECT name FROM qiita.software WHERE software_id = %s"
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchlast()
+
+    @property
+    def version(self):
+        """The version of the software
+
+        Returns
+        -------
+        str
+            The version of the software
+        """
+        with TRN:
+            sql = "SELECT version FROM qiita.software WHERE software_id = %s"
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchlast()
+
+    @property
+    def description(self):
+        """The description of the software
+
+        Returns
+        -------
+        str
+            The software description
+        """
+        with TRN:
+            sql = """SELECT description
+                     FROM qiita.software
+                     WHERE software_id = %s"""
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchlast()
+
+    @property
+    def commands(self):
+        """The list of commands attached to this software
+
+        Returns
+        -------
+        list of int
+            The command identifiers
+        """
+        with TRN:
+            sql = """SELECT command_id
+                     FROM qiita.software_command
+                     WHERE software_id = %s"""
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchflatten()
+
+    @property
+    def publications(self):
+        """The publications attached to the software
+
+        Returns
+        -------
+        list of 2-tuples of str
+            The list of DOI and pubmed_id attached to the publication
+        """
+        with TRN:
+            sql = """SELECT p.doi, p.pubmed_id
+                        FROM qiita.publication p
+                            JOIN qiita.software_publication sp
+                                ON p.doi = sp.publication_doi
+                        WHERE sp.software_id = %s"""
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchindex()
+
+    def add_publications(self, publications):
+        """Add publications to the software
+
+        Parameters
+        ----------
+        publications : list of 2-tuples of str
+            A list with the (DOI, pubmed_id) of the publications to be attached
```

Comment:
Maybe worth adding a link to what a PubMedID is? https://www.nlm.nih.gov/bsd/disted/pubmedtutorial/020_830.html